internal/server/add_payment.go

```diff
@@ -0,0 +1,83 @@
+package server
+
+import (
+	"github.com/ministryofjustice/opg-go-common/template"
+	"github.com/ministryofjustice/opg-sirius-lpa-frontend/internal/sirius"
+	"net/http"
+	"regexp"
+	"strconv"
+)
+
+type AddPaymentClient interface {
+	AddPayment(ctx sirius.Context, caseID int, amount int, source string, paymentDate sirius.DateString) error
+	Case(sirius.Context, int) (sirius.Case, error)
+}
+
+type addPaymentData struct {
+	XSRFToken string
+	Success   bool
+	Error     sirius.ValidationError
+
+	Case        sirius.Case
+	Amount      string
+	Source      string
+	PaymentDate sirius.DateString
+}
+
+func AddPayment(client AddPaymentClient, tmpl template.Template) Handler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		caseID, err := strconv.Atoi(r.FormValue("id"))
+		if err != nil {
+			return err
+		}
+
+		ctx := getContext(r)
+		data := addPaymentData{
+			XSRFToken:   ctx.XSRFToken,
+			Amount:      postFormString(r, "amount"),
+			Source:      postFormString(r, "source"),
+			PaymentDate: postFormDateString(r, "paymentDate"),
+		}
+
+		data.Case, err = client.Case(ctx, caseID)
+		if err != nil {
+			return err
+		}
+
+		if r.Method == http.MethodPost {
+			m, err := regexp.Match(`^\d*\.\d{2}$`, []byte(data.Amount))
```

Comment:
You probably want `\d+` rather than `\d*` to ensure that there's at least one digit before the decimal (otherwise `.45`) would be allowed
```suggestion
			m, err := regexp.Match(`^\d+\.\d{2}$`, []byte(data.Amount))
```

Reply:
Well spotted!